repl: don't use tty control codes when $TERM is set to "dumb"

saljam opened the pull request:
This stops the REPL from using ANSI control codes on terminals
which don't support them. (Fixes nodejs/node-v0.x-archive#5344)

For history see nodejs/node-v0.x-archive#25506

silverwind commented:
@saljam sorry about this, but could rebase this PR once against current master?

saljam commented:
Eh, I'm a little confused about what just happened. But sure, I just rebased. :)

silverwind commented:
We had a happy little issue (https://github.com/nodejs/node/issues/2713) which led us to force-pushing a few commits on top of master, and as your PR was based on one of the commits that was gone, it could no longer apply. The Github UI bugging out is another result of that.

Also, CI of this PR here: https://ci.nodejs.org/job/node-test-pull-request/258/

silverwind commented:
I think this is pretty low-risk, but could you add a test?

Fishrock123 commented:
I can foresee a problem here. `terminal` controls a bunch of different things, and not just ANSI control codes.

saljam commented:
@Fishrock123 as far as I can tell, it also disables readline and the up-down-arrow history.

Do we want readline on dumb terminals? (I seem to remember there being a reason I didn't just go with `useColors`, but that was over two years ago and I can't see why not now...)

saljam commented:
OK, I've switched to disabling `useColor` when `TERM=dumb` and added a basic test.

Fishrock123 commented:
LGTM. @saljam could you put a link into the commit log, and/or a comment in the code, to a reference on `env.TERM`? (i.e. if there is any sort of spec that defines it or something.)

saljam commented:
@Fishrock123 done. PTAL

Git, hg, grep, &c all respect TERM=dumb, but surprisingly I couldn't find a clear definition other than in terminfo's source.

silverwind commented:
LGTM

silverwind commented:
Landed with a few fixes to the test in https://github.com/nodejs/node/commit/ccea33df6262e82503942dfdbe45e7baade89caa.

I added a slightly more helpful error message to the asserts and fixed the linter errors. @saljam please run `make jslint` next time before committing ;)

saljam commented:
Thanks! And noted, somehow I convinced myself `make test` included a jslint
run.

On Tue, 22 Sep 2015 20:44 silverwind <notifications@github.com> wrote:

> Landed with a few fixes to the test in ccea33d
> <https://github.com/nodejs/node/commit/ccea33df6262e82503942dfdbe45e7baade89caa>
> .
>
> I added a slightly more helpful error message to the asserts and fixed the
> linter errors. @saljam <https://github.com/saljam> please run make jslint
> next time before committing ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/2712#issuecomment-142360112>.
>

silverwind commented:
@saljam it [does](https://github.com/nodejs/node/blob/ccea33df6262e82503942dfdbe45e7baade89caa/Makefile#L89-L92) include linting, and should've shown these errors after the test run.

saljam commented:
In which case whoops! Sorry! (I can confirm it does work.)

On Tue, 22 Sep 2015 20:52 silverwind <notifications@github.com> wrote:

> @saljam <https://github.com/saljam> it does
> <https://github.com/nodejs/node/blob/ccea33df6262e82503942dfdbe45e7baade89caa/Makefile#L89-L92>
> include linting, and should've shown these errors after the test run.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/2712#issuecomment-142363757>.
>
